### DISCUSSION TITLE
How to apply a custom function on a dataset's chunks to reduce its size and save it on disk ?

### DISCUSSION THREAD

**Discussion by vdevauxchupin** [pydata/xarray#7913] [Q&A] [answered]
Hello,

This question is a clone of something I asked on [StackOverflow](https://stackoverflow.com/questions/76394194/xarray-write-large-dataset-on-memory-without-killing-the-kernel). 

I have a dataset (_110 Gb_) that is created by using `combine_first` on a list gathering several datasets (saved as `zarr`) on a remote server.
This dataset is chunked and too big for my RAM. I want to write it on my local disk, but I overload my RAM whatever the method I use (I have 96Gb of RAM). 

**In short:** how can I write on disk a dataset that is too big to be stored in the RAM ?

**Why am I doing that ?** I want to apply a computationally intensive method that goes iteratively through every 2D pixel of my dataset (`x,y`) and works with their entire time dimension (`mid_date`). I thought the best way of doing that was to first download the dataset, then work on it locally. Otherwise accessing the pixels one by one takes too much time (100000 pixels, 5 seconds/pixel access). I also asked a question on StackOverflow concerning that, but got no satisfying answer [https://stackoverflow.com/questions/76384679/chunked-xarray-load-only-1-cell-in-memory-efficiently].

###############################################################

_Here is a code reproducing what I am trying to do (also as a python script):_

Dataset: ![dDYrQ](https://github.com/pydata/xarray/assets/73993845/ca62b3b0-1b25-4bb9-8459-ee7f841ebee6)

```

# import the necessary packages
import xarray as xr
import os
import sys
from dask.diagnostics import ProgressBar
import numpy as np

# Define some temporal boundaries
sdate = '2018-01-01'
edate = '2023-06-02'

# URLs of the datasets I want to merge as one
cubes = ['http://its-live-data.s3.amazonaws.com/datacubes/v02/N60W140/ITS_LIVE_vel_EPSG3413_G0120_X-3250000_Y350000.zarr',
 'http://its-live-data.s3.amazonaws.com/datacubes/v02/N50W130/ITS_LIVE_vel_EPSG3413_G0120_X-3350000_Y250000.zarr',
 'http://its-live-data.s3.amazonaws.com/datacubes/v02/N60W130/ITS_LIVE_vel_EPSG3413_G0120_X-3250000_Y250000.zarr',
 'http://its-live-data.s3.amazonaws.com/datacubes/v02/N50W140/ITS_LIVE_vel_EPSG3413_G0120_X-3350000_Y350000.zarr']

# Define spatial boundaries so you don't end up downloading tens of terabytes
xmin =-3340199.841439601
xmax =-3266124.863581888
ymin =273869.3657119706
ymax =364857.29614944384


# Create path to the save the merged dataset
pathsave = 'Datasets/'
# Create a subfolder
os.makedirs(pathsave, exist_ok = True)
# Here we drop all the variables except from those ones otherwise the cube is even larger
variables_to_keep = ['v', 'date_dt', 'acquisition_date_img1', 'acquisition_date_img2']

# Function to load the datasets
def load_datacubes(datacube_address, sdate, edate, xmax, xmin, ymax, ymin, variables_drop):

    # Grab the time values
    t = xr.open_dataset(datacube_address, engine='zarr').mid_date.values

    # Create a time mask, based on the validity of layers and the custom date-range
    t_mask = np.logical_and(t>np.datetime64(sdate), t<np.datetime64(edate))

    # Load datacube according to prerequisites (time, space and variables)
    ds = xr.open_zarr(datacube_address,
                chunks=({'mid_date': 3000, 'y': 300, 'x': 300}),
                drop_variables=variables_drop
                ).sel(  mid_date = t_mask,
                        x=slice(xmin, xmax),
                        y=slice(ymax, ymin)).drop_duplicates(dim='mid_date')
    ds.attrs={}
    return ds

# Add the dimensions to the download otherwise they are dropped
variables_to_keep += ['mid_date', 'x', 'y']

# List of variables to drop for the download (we drop everything but the variables in "variables to keep")
variables_drop = [ele for ele in list(
        xr.open_dataset(cubes[0], engine='zarr').variables
        ) if ele not in variables_to_keep
]

# Load the cubes, chunked
datacubes = [load_datacubes(datacube_address, sdate, edate, xmax, xmin, ymax, ymin, variables_drop).sortby(['mid_date','y','x']).chunk({'mid_date': 3000, 'y': 300, 'x': 300}) for datacube_address in cubes]


# Combine the datacubes into one (stitches the cubes spatially, and aligns them temporally when they have the same timestamps)
xrds = datacubes[0]
xrds = xrds.combine_first(datacubes[1])
xrds = xrds.combine_first(datacubes[2])
xrds = xrds.combine_first(datacubes[3])

#Option 1, did not work
#write_job = xrds.to_zarr(f"{pathsave}Test.zarr", mode='w', compute=False, consolidated=True)

#Option 2, did not work (with chunk size {'mid_date':3000, 'x':100, 'y':100})


#write_job = xrds.to_netcdf(f"test.nc",compute=False)
#with ProgressBar():
#    print(f"Writing to {pathsave}")
#    write_job = write_job.compute()                 

# Option 3, did not work. That's the option I took the screenshot from
# I force the chunks to be really small so I don't overload the memory
chunk_size = {'mid_date':1, 'y':xrds.y.shape[0], 'x':xrds.x.shape[0]}

with ProgressBar():
    for i, (key, chunk) in enumerate(xrds.chunk(chunk_size).items()):
        chunk_dataset = xr.Dataset({key: chunk})
        chunk_dataset.to_netcdf(f"{pathsave}/chunk_{i}.nc", mode="w", compute=True)
        print(f"Chunk {i+1} saved.")    
```

**Comment by dcherian** [comment]
Does any of the material in https://e-marshall.github.io/itslive/intro.html help?

cc @e-marshall

**Reply by e-marshall** [reply]
The tutorial didn't go into anything that required loading into memory IIRC

**Comment by e-marshall** [comment]
Hey @vdevauxchupin ! I've struggled with this type of thing of thing with ITSLIVE before but don't think I've figured out much of a solution, and haven't figured out a way to do directly what you're asking. A few thoughts: 
- can you reduce the spatial size / x,y dimensions of the object any further ? (most of the glaciers I've been working with are relatively small which has worked to my advantage at times ) . In the past I have called `compute()` on an object (smaller than the one you're working with) right before a non-lazy operation and then `del object` right after and that seems to work. it hasn't been cleaned but there's an example of that in the `def calc_glacier_medians_1granule` function in this [notebook](https://github.com/e-marshall/itslive_velocity/blob/master/2022/agu_22_processing/itslive_screener_32645.ipynb) (all the way at the bottom)
- what operation do you want to apply to each 2d xy array of the object? Is it one that could be implemented in a vectorized way with [`apply_ufunc`](https://docs.xarray.dev/en/stable/examples/apply_ufunc_vectorize_1d.html), [map_blocks()](https://docs.xarray.dev/en/stable/generated/xarray.map_blocks.html) or [reduce()](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.reduce.html)

**Reply by vdevauxchupin** [reply]
Hi @e-marshall  ! Thanks for your answer.

There are several ways I sometimes reduce the dataset size:

- Use the quality index of each slice (amount of NaNs per slice), and kick out every slice below a quality threshold. CON: you get rid of some potentially interesting data

- Use a glacier mask to download only the pixels on the ice. CON: annoying to automatize and you need yet another dataset for that (but it seems you've managed to do that ?)

- Invert a velocity function (to be applied on each 2d xy array) that takes what you have, but can be resampled in order to get 1 value per day (or 1 value per 5 days, or whatever you want). CON: takes a while to compute (9h for 11Gb of data, non-parallelized). But that function is the one that reduces the dataset's size the best way while ensuring you don't mix yearly or seasonal signals..

- Use the `resample` capabilities of xarray to force 1 value per day or less. CON: the result is wrong, as the velocities are generated from different time intervals (you end up mixing seasonal and yearly signals altogether). The outcome is the same as the inverse function, only this method is not scientifically correct.


I did find a way to make it work somehow:

- Not use `combine_first()`, but instead save every chunked dataset individually, apply the inverse function, and then combine the smaller datasets together. CON: Will take approximately 4 days of computation, unless I manage to parallelize my inverse function.


But in the end the issues remain the same: if you don't have a system with a lot of RAM (at least 64Gb), this becomes a complicated issue.

**Reply by e-marshall** [reply]
nice! let's add this to the list of things to talk about when we meet -  this seems like an unavoidable step in these workflows so would be great to brainstorm together some. the inverse function seems like a neat approach. i've mostly been using the glacier mask approach and wrapping clipping by a glacier outline with some other processing steps, but its also very slow...

**Reply by vdevauxchupin** [reply]
Oh I see ! Sometimes I wish matrices could have other shapes than rectangles.
I let the use clip the dataset to their AOI with the mouse on the GUI, it might not be as precise as the glacier limits but it shouldn't change much in the end. 
However the inversion could definitely be sped-up if it is applied to the glacier pixels only, instead of everything on the image

**Comment by TomNicholas** [comment, marked as answer]
> I want to apply a computationally intensive method that goes iteratively through every 2D pixel of my dataset (x,y) and works with their entire time dimension (mid_date). I thought the best way of doing that was to first download the dataset, then work on it locally.

It sounds more like you should do this:
- Open the dataset lazily with a chunking scheme like `{'x': 100, 'y': 100. 'time': -1}` (so same as chunks on disk expect for opening all time chunks as contiguous). 
- Then you want to write a function that can happily apply your analysis to a single chunk after loading just that chunk into memory. Because you want to apply elementwise (to each pixel) you'll need to look into what np.vectorize does.
- Then one that works for one chunk you can worry about using it on many chunks in one go. The function you wrote in the last step can be applied with `xr.apply_ufunc`, and now you can use `dask='parallelized'` to avoid the bigger than RAM issue. Of course this assumes that you didn't just try to download the whole thing locally. Do you have access to a jupyterhub or similar where you could run this using dask?

**Reply by vdevauxchupin** [reply]
Thanks for your reply ! Yes I have access to a JupyterHub, it doesn't have yet a Dask Gateway implemented. But the machine I work with has Dask installed on it. 

I tried unsuccessfully to use `xr.apply_ufunc` and `map_blocks`. The challenges to me are that:

- My function takes an input vector of length N, and returns a smaller output vector of size P where P<N
- If I work with the chunked dataset directly (`ds`), applying the function is non-trivial because the outputs have to be stored in an array that has a smaller size than `ds` on the `mid_date` axis only. 
- If I work with the array inside the dataset directly (`ds.v.values`), parallelizing the function is slower than computing in loops because each core has to constantly talk to the others to store the function's outputs in the output array correctly.



The only way I managed to make that work (parallelization worked but was slower):

```
# Function that takes a cell of the dataset and reduces its size
def foo(vector, output_array,i,j):
    # Perform some operations on the vector
    # and return a smaller result

    # Example: Take the sum of every 10 elements
    output_array[:,i,j] = np.sum(vector.reshape(-1, 10), axis=1)
    
# Pre-determine the size of the dataset and the output_array
N = 1000
P = 100

# Simulates the variable of my dataset that I want to iterate on (equivalent to ds.v.values)
ds_var = np.random.rand((N,300,300))

# Pre-allocate the output array
output_array =  np.zeros((P, ds_var.shape[1] , ds_var.shape[2]))

# Run the function and store the results in output_array
for i in range(ds_var.shape[1]):
        for j in range(ds_var.shape[2]):
                  foo(ds_var[:,i,j], output_array, i, j)

```

Sorry if it's not clear, it's challenging to explain.

**Reply by vdevauxchupin** [reply]
I managed to make it work, benchmark shows advantage for the for-loop method that weakens as the dataset's size increases: 
```

import numpy as np
import xarray as xr
import time

# Function applied on the dataset
def foo(vector):
    # Perform some operations on the vector
    # and return a smaller result

    # Example: Take the sum of every 10 elements
    return np.sum(vector.reshape(-1, 10), axis=1)

# Function applied in loop
def foo2(vector, output, i, j):
    # Perform some operations on the vector
    # and return a smaller result

    # Example: Take the sum of every 10 elements
    output[:,i,j] = np.sum(vector.reshape(-1, 10), axis=1)

# Set the time dimension
N = 1000
P = 100

# Create the input data
date = np.arange(N)
y = np.arange(334)
x = np.arange(333)
input_data = np.random.rand(len(date), len(y), len(x))

# Create the input DataArray
input_da = xr.DataArray(input_data, dims=['date', 'y', 'x'],
                        coords={'date': date, 'y': y, 'x': x}).chunk({'date':-1, 'y': 100, 'x': 100})

###### WITH APPLY_UFUNC
start_time = time.time()
# Apply the function using apply_ufunc
output_data = xr.apply_ufunc(foo, input_da.transpose('date', 'y', 'x'),
                             vectorize=True,
                             input_core_dims=[['date']],
                             output_core_dims=[['date2']],
                             dask='parallelized',
                             output_dtypes=[float],
                             dask_gufunc_kwargs={'output_sizes': {'date2': P}}
                             )
output_data.compute()
# Stop the timer and calculate the elapsed time
end_time = time.time()
elapsed_time = end_time - start_time

# Print the elapsed time
print("Elapsed time:", elapsed_time, "seconds")



###### WITH FOR LOOPS

#Pre-allocate array
output = np.zeros((P,334,333))

# Start timing
start_time = time.time()

# Load in memory the dataarray's array
dsval = input_da.values
for i in range(input_da.shape[1]):
    for j in range(input_da.shape[2]):
        foo2(dsval[:,i,j], output, i, j)
        
# Stop the timer and calculate the elapsed time
end_time = time.time()
elapsed_time = end_time - start_time

# Print the elapsed time
print("Elapsed time:", elapsed_time, "seconds")
```

Elapsed time: 7.2942163944244385 seconds (apply_ufuncs)
Elapsed time: 1.8031437397003174 seconds (for-loop)

**Reply by TomNicholas** [reply]
> I managed to make it work

Nice! That looks much more like what I was imagining.

> input_da.transpose('date', 'y', 'x'),

If you make your function `foo` act on the last axis (so `axis=-1`, or wrap it in `np.apply_along_axis` or similar) then you shouldn't need to do any transposition as `apply_ufunc` will handle that for you.

You should read this [tutorial documentation on `apply_ufunc`](https://tutorial.xarray.dev/advanced/apply_ufunc/apply_ufunc.html) too if you haven't already.

> benchmark shows advantage for the for-loop method that weakens as the dataset's size increases:

I would expect using xarray for this to be slower for small data than not using xarray, what we're trying to do is write code that can scale to larger-than-memory datasets.

**Reply by vdevauxchupin** [reply]
> You should read this [tutorial documentation on apply_ufunc](https://tutorial.xarray.dev/advanced/apply_ufunc/apply_ufunc.html) too if you haven't already.

Thanks for the link ! I will work on these examples.

> I would expect using xarray for this to be slower for small data than not using xarray, what we're trying to do is write code that can scale to larger-than-memory datasets.

That was my primary goal and it seems to be working. I'm working with my original dataset and the memory looks stable. I was wondering if the parallelization would speed things up though, because my `foo` function could definitely be sped up while parallelizing, I just haven't figured it out (I never really parallelized before).

**Reply by TomNicholas** [reply]
Dask should parallelize for you over chunks. So in your case it could parallelize over x & y chunks, but not over time chunks, nor over individual pixels within those x and y chunks. (Parallelizing over those pixels too could be done, but I wouldn't worry about that yet.)

You should also read the [main documentation page on using dask with xarray](https://docs.xarray.dev/en/stable/user-guide/dask.html).

> Yes I have access to a JupyterHub, it doesn't have yet a Dask Gateway implemented. But the machine I work with has Dask installed on it.

Using a `LocalCluster` will be sufficient to analyse data larger than memory and to get some parallelism. You shouldn't need to use `dask_gateway` unless you want parallelism across many many workers.

**Reply by vdevauxchupin** [reply]
> If you make your function foo act on the last axis (so axis=-1, or wrap it in np.apply_along_axis or similar) then you shouldn't need to do any transposition as apply_ufunc will handle that for you.

I realized that the transpose isn't even necessary. However, do you think this might be faster if I flatten both input and output datasets ? Or it won't change anything ?

**Reply by vdevauxchupin** [reply]
> Dask should parallelize for you over chunks. So in your case it could parallelize over x & y chunks, but not over time chunks, nor over individual pixels within those x and y chunks. (Parallelizing over those pixels too could be done, but I wouldn't worry about that yet.)

Oh I thought chunking `x` and `y`, and applying `ufuncs` with `dask='Paralellized' was already computing in parallel for me in that case. I will look into it, thanks !

**Reply by TomNicholas** [reply]
> flatten both input and output datasets

I don't think it will change anything.

**Reply by TomNicholas** [reply]
> was already computing in parallel for me 

It should be, but only over dask chunks, not within them.

**Reply by dcherian** [reply]
Here's a [WIP preview of reworked apply_ufunc tutorial material](https://xarray-contrib-xarray-tutorial-preview-pr-180.surge.sh/advanced/apply_ufunc/apply_ufunc.html) just in case, though it looks like you got things working

**Reply by vdevauxchupin** [reply]
> It should be, but only over dask chunks, not within them.
Ok I see, thank you ! 

> Here's a [WIP preview of reworked apply_ufunc tutorial material](https://xarray-contrib-xarray-tutorial-preview-pr-180.surge.sh/advanced/apply_ufunc/apply_ufunc.html) just in case, though it looks like you got things working

Thank you for the link ! It;s super useful. Without tutorials I have trouble understanding otherwise.

**Reply by TomNicholas** [reply]
Great. Has your original question been answered now? If not you can always open another discussion for any follow-up questions too.

**Reply by vdevauxchupin** [reply]
I will close this question but maybe modify the title since the original question hasn't really been answered: how to write on disk a dataset too big for the RAM ?

**Reply by TomNicholas** [reply]
To answer your original question you should be able to follow the same strategy that you now are using  to process the data chunk-by-chunk to instead write the data chunk-by-chunk. Basically instead of applying your analysis function you use `to_zarr` or `to_netcdf` instead, and run it with dask.

**Reply by vdevauxchupin** [reply]
Ok I will try that. Thank you for helping me through that !
